pallets/messages/src/migration/v2/mod.rs

```diff
@@ -0,0 +1,54 @@
+use crate::{Config, MessageIndex, Pallet};
+use common_primitives::{msa::MessageSourceId, schema::SchemaId};
+use core::fmt::Debug;
+use frame_support::{pallet_prelude::*, storage_alias};
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_runtime::codec::{Decode, Encode};
+
+// Type aliases for readability
+#[allow(type_alias_bounds)]
+pub(crate) type K1Type<T: Config> = BlockNumberFor<T>;
+#[allow(type_alias_bounds)]
+pub(crate) type K2Type = SchemaId;
+#[allow(type_alias_bounds)]
+pub(crate) type K3Type = MessageIndex;
+
+/// Storage for messages in v2 and lower
+/// - Key: (block_number, schema_id, message_index)
+/// - Value: Message
+#[storage_alias]
+pub(crate) type MessagesV2<T: Config> = StorageNMap<
+	Pallet<T>,
+	(
+		storage::Key<Twox64Concat, K1Type<T>>,
+		storage::Key<Twox64Concat, K2Type>,
+		storage::Key<Twox64Concat, K3Type>,
+	),
+	Message<<T as Config>::MessagesMaxPayloadSizeBytes>,
+	OptionQuery,
+>;
+
+/// Ephemeral storage key for tracking the completion status of the migration
+/// in order to perform the final step and for try-runtime. MUST be killed at
+/// the end of the migration!
+#[storage_alias]
+pub(crate) type DoneV3Migration<T: Config> = StorageValue<Pallet<T>, bool, ValueQuery>;
```

Comment:
nit: Shouldn't this be part of the Storage of the pallet? `storage_alias` is usually used to access existing storages with old type deserialization. In this case this is a brand new storage and might make sense to just move it into the pallet even though maybe `storage_alias` can be used to achieve the same.

Reply:
The only real difference between a storage alias and "regular" storage is that the key prefix is registered in the pallet & exposed in metadata.

Interestingly, this could be a way to define "private" storage inside a pallet that's "hidden" from the metadata (though you could still retrieve it by hex key, if you compute the hash from the source).
Not sure what that buys anyone, aside from discouraging relying on the storage API vs runtime API for queries... but interesting.